jax/lax/lax.py

```diff
@@ -3831,15 +3831,137 @@ def _gather_dtype_rule(operand, start_indices, **kwargs):
 
 def _gather_shape_rule(operand, start_indices, *, dimension_numbers,
                        slice_sizes):
-  if len(operand.shape) != len(slice_sizes):
-    msg = ("slice_sizes must have rank equal to the gather operand; "
-          "operand.shape={}, slice_sizes={}".format(operand.shape, slice_sizes))
-    raise ValueError(msg)
-  result_rank = len(dimension_numbers.offset_dims) + start_indices.ndim - 1
-  start_indices_shape = iter(start_indices.shape[:-1])
-  slice_sizes = iter(np.delete(slice_sizes, dimension_numbers.collapsed_slice_dims))
-  return tuple(next(slice_sizes) if i in dimension_numbers.offset_dims
-               else next(start_indices_shape) for i in range(result_rank))
+  """Validates the well-formedness of the arguments to Gather.
+
+  The code implements the checks based on the detailed operation semantics of
+  XLA's `Gather <https://www.tensorflow.org/xla/operation_semantics#gather>`_
+  operator and following the outline of the implementation of
+  ShapeInference::InferGatherShape in TensorFlow.
+  """
+  rank = lambda arr: len(arr.shape)
+
+  def _is_sorted(dims, name):
```

Comment:
It seems that some of these helper functions can be shared with the _scatter_shape_rule, right? Pulling them out and sharing them may make the actual shape rules easier to read.

Reply:
Addressed in `c2907b2b`!